File: component-acgnews/src/main/java/com/rabtman/acgnews/mvp/model/ZeroFiveNewsModel.java
package com.rabtman.acgnews.mvp.model;

import com.fcannizzaro.jsoup.annotations.JP;
import com.rabtman.acgnews.mvp.contract.ZeroFiveNewsContract;
import com.rabtman.acgnews.mvp.model.jsoup.ZeroFiveNewsPage;
import com.rabtman.common.base.mvp.BaseModel;
import com.rabtman.common.di.scope.FragmentScope;
import com.rabtman.common.integration.IRepositoryManager;
import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.FlowableEmitter;
import io.reactivex.FlowableOnSubscribe;
import io.reactivex.annotations.NonNull;
import javax.inject.Inject;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;

/**
 * @author Rabtman
 */
@FragmentScope
public class ZeroFiveNewsModel extends BaseModel implements ZeroFiveNewsContract.Model {

  @Inject
  public ZeroFiveNewsModel(IRepositoryManager repositoryManager) {
    super(repositoryManager);
  }

  @Override
  public Flowable<ZeroFiveNewsPage> getAcgNews(final String typeUrl) {
    return Flowable.create(new FlowableOnSubscribe<ZeroFiveNewsPage>() {
      @Override
      public void subscribe(@NonNull FlowableEmitter<ZeroFiveNewsPage> e) throws Exception {
        Element html = Jsoup.connect(typeUrl).get();
        if(html == null){
          e.onError(new Throwable("element html is null"));
        }else {
          ZeroFiveNewsPage zeroFiveNewsPage = JP.from(html, ZeroFiveNewsPage.class);
          e.onNext(zeroFiveNewsPage);
          e.onComplete();
        }
      }
    }, BackpressureStrategy.BUFFER);
  }

}

File: component-acgnews/src/main/java/com/rabtman/acgnews/mvp/model/ZeroFiveNewsModel.kt
package com.rabtman.acgnews.mvp.model

import com.fcannizzaro.jsoup.annotations.JP
import com.rabtman.acgnews.mvp.contract.ZeroFiveNewsContract
import com.rabtman.acgnews.mvp.model.jsoup.ZeroFiveNewsPage
import com.rabtman.common.base.mvp.BaseModel
import com.rabtman.common.di.scope.FragmentScope
import com.rabtman.common.integration.IRepositoryManager
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import javax.inject.Inject

/**
 * @author Rabtman
 */
@FragmentScope
class ZeroFiveNewsModel @Inject constructor(repositoryManager: IRepositoryManager?) : BaseModel(repositoryManager), ZeroFiveNewsContract.Model {
    override fun getAcgNews(typeUrl: String?): Flowable<ZeroFiveNewsPage> {
        return Flowable.create({ e ->
            val html: Element? = Jsoup.connect(typeUrl).get()
            if (html == null) {
                e.onError(Throwable("element html is null"))
            } else {
                val zeroFiveNewsPage = JP.from(html, ZeroFiveNewsPage::class.java)
                e.onNext(zeroFiveNewsPage)
                e.onComplete()
            }
        }, BackpressureStrategy.BUFFER)
    }
}